### PR TITLE
new_menu_helpers, multiboot, hof_pc, play_time, decompress, reset_save_heap, save_location, mail_data, fldeff_*

### PR DESCRIPTION
**Pull request by jiangzhengwenjz**

### CODE REVIEW

File: src/new_menu_helpers.c [hunk 1]
```diff
@@ -0,0 +1,765 @@
...
+        gTasks[taskId].data[0] = CopyDecompressedTileDataToVram(bgId, ptr, size, offset, mode);
+        SetWordTaskArg(taskId, 1, (u32)ptr);
+    }
+}
+
+void DecompressAndLoadBgGfxUsingHeap2(u8 bgId, const void *src, u32 size, u16 offset, u8 mode)
+{
+    u32 sizeOut;
+    void *ptr = MallocAndDecompress(src, &sizeOut);
+        if (sizeOut > size)
```
Comment (garakmon) [inline src/new_menu_helpers.c]: un-tab

File: src/new_menu_helpers.c [hunk 2]
```diff
@@ -0,0 +1,765 @@
...
+#include "script.h"
+#include "graphics.h"
+#include "palette.h"
+
+#define DLG_WINDOW_PALETTE_NUM 15
+#define DLG_WINDOW_BASE_TILE_NUM 0x200
+#define STD_WINDOW_PALETTE_NUM 14
+#define STD_WINDOW_BASE_TILE_NUM 0x214
+
+static EWRAM_DATA bool8 gUnknown_203AB58[4] = {FALSE}; // knizz: bgmaps_that_need_syncing
```
Comment (garakmon) [inline src/new_menu_helpers.c]: what does this comment mean?

Comment (jiangzhengwenjz) [reply src/new_menu_helpers.c]: replied in discord

File: src/new_menu_helpers.c [hunk 3]
```diff
@@ -0,0 +1,765 @@
...
+static EWRAM_DATA u16 gUnknown_203AB5C = {0};
+static EWRAM_DATA void *gUnknown_203AB60[0x20] = {NULL};
+static EWRAM_DATA u8 sStartMenuWindowId = {0};
+
+static const u8 gUnknown_841F428[] = { 8, 4, 1, 0, };
```
Comment (PikalaxALT) [inline src/new_menu_helpers.c]: Last element in an array or struct should not be followed by a comma

File: src/new_menu_helpers.c [hunk 4]
```diff
@@ -0,0 +1,765 @@
...
+    {
+        CopyBgTilemapBufferToVram(3);
+        gUnknown_203AB58[3] = FALSE;
+    }
+}
+
+void ResetTempTileDataBuffers(void)
+{
+    int i;
+    for (i = 0; i < (s32)ARRAY_COUNT(gUnknown_203AB60); i++)
```
Comment (PikalaxALT) [inline src/new_menu_helpers.c]: Use `NELEMS` for consistency

File: src/new_menu_helpers.c [hunk 5]
```diff
@@ -0,0 +1,765 @@
...
+        gTasks[taskId].data[0] = CopyDecompressedTileDataToVram(bgId, ptr, sizeOut, offset, mode);
+        SetWordTaskArg(taskId, 1, (u32)ptr);
+    }
+}
+
+static void TaskFreeBufAfterCopyingTileDataToVram(u8 taskId)
+{
+    if (!CheckForSpaceForDma3Request(gTasks[taskId].data[0]))
+    {
+        Free((void *)GetWordTaskArg(taskId, 1));
```
Comment (PikalaxALT) [inline src/new_menu_helpers.c]: No need to cast to `void *`

File: src/new_menu_helpers.c [hunk 6]
```diff
@@ -0,0 +1,765 @@
...
+        DestroyTask(taskId);
+    }
+}
+
+void *MallocAndDecompress(const void *src, u32 *size)
+{
+    void *ptr;
+    u8 *sizeAsBytes = (u8 *)size;
+    u8 *srcAsBytes = (u8 *)src;
```
Comment (PikalaxALT) [inline src/new_menu_helpers.c]: No need to typecast a void pointer when assigning it to a non-void pointer variable

File: src/new_menu_helpers.c [hunk 7]
```diff
@@ -0,0 +1,765 @@
...
+}
+
+void AddTextPrinterDiffStyle(bool8 allowSkippingDelayWithButtonPress)
+{
+    u8 result;
+    void *nptr = NULL; // This is required for matching
+
+    gTextFlags.canABSpeedUpPrint = allowSkippingDelayWithButtonPress;    
+    result = ContextNpcGetTextColor();
+    if (!result)
```
Comment (PikalaxALT) [inline src/new_menu_helpers.c]: In this case, `if (result == 0)` is preferred for consistency with line 435

File: src/new_menu_helpers.c (hunk 5 above)
Comment (jiangzhengwenjz) [reply src/new_menu_helpers.c]: It's actually required, because the return value is not pointer

Comment (PikalaxALT) [reply src/new_menu_helpers.c]: Oh, my mistake.

File: src/new_menu_helpers.c (hunk 6 above)
Comment (jiangzhengwenjz) [reply src/new_menu_helpers.c]: The problem lying here is that I'm assigning a const ptr to a non-const one, should I modify function signature?

File: src/new_menu_helpers.c (hunk 3 above)
Comment (PikalaxALT) [reply src/new_menu_helpers.c]: Also, the last element (0) can be removed from the definition as it is merely padding in this context

File: src/new_menu_helpers.c (hunk 6 above)
Comment (PikalaxALT) [reply src/new_menu_helpers.c]: No, just declare `srcAsBytes` as `const u8 *`

File: src/new_menu_helpers.c [hunk 8]
```diff
@@ -0,0 +1,765 @@
...
+    ptr = Alloc(*size);
+    if (ptr)
+        LZ77UnCompWram(src, ptr);
+    return ptr;
+}
+
+static u16 CopyDecompressedTileDataToVram(u8 bgId, const void *src, u16 size, u16 offset, u8 mode)
+{
+    switch (mode)
+    { // different to EM
```
Comment (PikalaxALT) [inline src/new_menu_helpers.c]: what does this comment mean?

Comment (jiangzhengwenjz) [reply src/new_menu_helpers.c]: The function logic got changed in EM. If you regard it as confusing I can delete all this kind of comment

Comment (PikalaxALT) [reply src/new_menu_helpers.c]: it's not necessary. if anything, the emerald version should have a "different from fr/lg" comment as that game released after frlg

File: src/reset_save_heap.c [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+#include "new_game.h"
+#include "overworld.h"
+#include "malloc.h"
+
+void sub_8079B7C(void)
+{
+    u16 imeBackup = REG_IME;
+    
+    REG_IME = 0;
+    RegisterRamReset(0x00000001);
```
Comment (PikalaxALT) [inline src/reset_save_heap.c]: `RegisterRamReset(RESET_EWRAM);`

File: tools/br_ips/br_ips.c [hunk 10]
```diff
@@ -50,6 +50,7 @@ static int getline(char ** lineptr, size_t * n, FILE * stream) {
     return i;
 }
 
+
```
Comment (PikalaxALT) [inline tools/br_ips/br_ips.c]: ????

File: src/bg.c [hunk 11]
```diff
@@ -928,79 +865,66 @@ void SetBgAffine(u8 bg, u32 srcCenterX, u32 srcCenterY, s16 dispCenterX, s16 dis
...
 }
 
-u8 Unused_AdjustBgMosaic(u8 a1, u8 a2)
-{
-    u16 result;
-    s16 test1;
-    s16 test2;
-
-    result = GetGpuReg(REG_OFFSET_MOSAIC);
+#define BG_MOSAIC_SET 0
```
Comment (jiangzhengwenjz) [inline src/bg.c]: forgot to mention previously... what style am I supposed to follow in this kind of situation? enum? and where to place it?

Comment (garakmon) [reply src/bg.c]: either is fine, but it should probably be in a header. especially if it is something that could be used in other places.

Comment (jiangzhengwenjz) [reply src/bg.c]: > either is fine, but it should probably be in a header (bg.h is fine). especially if it is something that could be used in other places.

Well, it's only referenced once (in poison field effect function, which was optimized in emerald to a direct IO buffer write). So should I put it in bg.h?

Comment (garakmon) [reply src/bg.c]: yeah, I would

File: .gitignore [hunk 12]
```diff
@@ -42,5 +42,6 @@ build/
 *.DS_Store
 *.pl
 *.bak
+src/data/items.h
 src/data/items.h # Autogenerated by jsonproc
```
Comment (PikalaxALT) [inline .gitignore]: Duplicate entry

File: src/battle_ai_script_commands.c [hunk 13]
```diff
@@ -970,331 +968,69 @@ static void BattleAICmd_get_move_power(void)
...
+    }
 
     if (gBattleMoves[AI_THINKING_STRUCT->moveConsidered].power > 1
-     && sDiscouragedPowerfulMoveEffects[i] == 0xFFFF)
+        && sDiscouragedPowerfulMoveEffects[i] == 0xFFFF)
     {
         gDynamicBasePower = 0;
-        eDynamicMoveType = 0;
-        eDmgMultiplier = 1;
+        *(&gBattleStruct->dynamicMoveType) = 0;
```
Comment (PikalaxALT) [inline src/battle_ai_script_commands.c]: Is this syntax necessary to match?

File: src/decompress.c [hunk 14]
```diff
@@ -0,0 +1,662 @@
...
+{
+    u8 *v1, *v2;
+    int i, j, k, l;
+
+    if (a0 & 1)
+    {
+        u8 v40 = ((a0 >> 1) + 4);
+        for (i = 0; i < a1; ++i)
+        {
+            // this nested loop failed to match
```
Comment (PikalaxALT) [inline src/decompress.c]: This right here is the final boss :P

File: src/decompress.c [hunk 15]
```diff
@@ -0,0 +1,662 @@
...
+        ");
+}
+#endif
+
+bool8 LoadCompressedSpriteSheetUsingHeap(const struct CompressedSpriteSheet* src)
+{
+    struct SpriteSheet dest;
+    void* buffer;
+
+    buffer = AllocZeroed(*((u32*)(&src->data[0])) >> 8);
```
Comment (PikalaxALT) [inline src/decompress.c]: Don't need `&src->data[0]` to create a pointer to the start of `src->data`.
Just use `src->data`.

Example:
```c
    buffer = AllocZeroed(*((u32 *)src->data) >> 8);
```

File: src/decompress.c [hunk 16]
```diff
@@ -0,0 +1,662 @@
...
+    Free(buffer);
+    return FALSE;
+}
+
+bool8 LoadCompressedSpritePaletteUsingHeap(const struct CompressedSpritePalette *src)
+{
+    struct SpritePalette dest;
+    void* buffer;
+
+    buffer = AllocZeroed(*((u32*)(&src->data[0])) >> 8);
```
Comment (PikalaxALT) [inline src/decompress.c]: Same here

File: src/fldeff_dig.c [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+    if (CanUseEscapeRopeOnCurrMap() == TRUE)
+    {
+        gFieldCallback2 = FieldCallback_PrepareFadeInFromMenu;
+        gPostMenuFieldCallback = hm2_dig;
+        return TRUE;
+    }
+    return FALSE;
+}
+
+static void hm2_dig(void)
```
Comment (PikalaxALT) [inline src/fldeff_dig.c]: Weird name, consider changing

File: src/fldeff_dig.c [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+    Overworld_ResetStateAfterDigEscRope();
+    FieldEffectStart(FLDEFF_USE_DIG);
+    gFieldEffectArguments[0] = GetCursorSelectionMonId();
+}
+
+bool8 FldEff_UseDig(void)
+{
+    u8 taskId = oei_task_add();
+
+    gTasks[taskId].data[8] = (u32)sub_80C9AFC >> 16;
```
Comment (PikalaxALT) [inline src/fldeff_dig.c]: The pointer storage here (and following line) is probably a macro expansion.

File: src/fldeff_rocksmash.c [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+        EventObjectSetGraphicsId(&gMapObjects[gPlayerAvatar.mapObjectId], GetPlayerAvatarGraphicsIdByCurrentState());
+        StartSpriteAnim(&gSprites[gPlayerAvatar.spriteId], gFieldEffectArguments[2]);
+        FieldEffectActiveListRemove(6);
+        gTasks[taskId].func = sub_80C99A0;
+    }
+}
+
+static void sub_80C99A0(u8 taskId)
+{
+    ((void (*)(void))(((u16)gTasks[taskId].data[8] << 16) | (u16)gTasks[taskId].data[9]))();
```
Comment (PikalaxALT) [inline src/fldeff_rocksmash.c]: This pointer read is probably a macro expansion

File: src/fldeff_strength.c [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+static void FldEff_UseStrength(void)
+{
+    gFieldEffectArguments[0] = GetCursorSelectionMonId();
+    ScriptContext1_SetupScript(EventScript_FldEffStrength);
+}
+
+bool8 sub_80D0860(void)
+{
+    u8 taskId = oei_task_add();
+    gTasks[taskId].data[8] = (u32)sub_80D08A8 >> 16;
```
Comment (PikalaxALT) [inline src/fldeff_strength.c]: See comment on fldeff_dig.c

File: src/fldeff_sweetscent.c [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+    gFieldEffectArguments[0] = GetCursorSelectionMonId();
+}
+
+bool8 FldEff_SweetScent(void)
+{
+    u8 taskId;
+
+    SetWeatherScreenFadeOut();
+    taskId = oei_task_add();
+    gTasks[taskId].data[8] = (u32)StartSweetScentFieldEffect >> 16;
```
Comment (PikalaxALT) [inline src/fldeff_sweetscent.c]: See comment on fldeff_dig.c

File: src/load_save.c [hunk 22]
```diff
@@ -132,28 +132,28 @@ void MoveSaveBlocks_ResetHeap(void)
 
 u32 sav2_x1_query_bit1(void)
 {
-    return gSaveBlock2Ptr->specialSaveWarp & 1;
+    return gSaveBlock2Ptr->specialSaveWarpFlags & 1;
```
Comment (PikalaxALT) [inline src/load_save.c]: Use the enums here and below

File: src/mail_data.c [hunk 23]
```diff
@@ -0,0 +1,182 @@
...
+    }
+    return 0xFF;
+}
+
+u16 SpeciesToMailSpecies(u16 species, u32 personality)
+{
+    return species == 0xC9 ? ((GetUnownLetterByPersonality(personality) << 16) + 0x75300000) >> 16 : species;
+}
+
+u16 MailSpeciesToSpecies(u16 mailSpecies, u16 *buffer)
```
Comment (PikalaxALT) [inline src/mail_data.c]: `buffer` should be `unownLetter`

File: src/mail_data.c [hunk 24]
```diff
@@ -0,0 +1,182 @@
...
+            gSaveBlock1Ptr->mail[id].itemId = itemId;
+            SetMonData(mon, MON_DATA_MAIL, &id);
+            SetMonData(mon, MON_DATA_HELD_ITEM, heldItem);
+            return id;
+        }
+    }
+    return 0xFF;
+}
+
+u16 SpeciesToMailSpecies(u16 species, u32 personality)
```
Comment (PikalaxALT) [inline src/mail_data.c]:
```c
u16 SpeciesToMailSpecies(u16 species, u32 personality)
{
    if (species == SPECIES_UNOWN)
    {
        species = GetUnownLetterByPersonality(personality);
        species += 1875;
    }
    return species;
}
```

File: src/mail_data.c [hunk 25]
```diff
@@ -0,0 +1,182 @@
...
+    u8 newHeldItem[2];
+
+    newHeldItem[0] = ITEM_NONE;
+    newHeldItem[1] = ITEM_NONE << 8;
+    newMailId = 0xFF;
+    for (i = PARTY_SIZE; i < MAIL_COUNT; i++)
+    {
+        if (gSaveBlock1Ptr->mail[i].itemId == ITEM_NONE)
+        {
+            memcpy(&gSaveBlock1Ptr->mail[i], &gSaveBlock1Ptr->mail[GetMonData(mon, MON_DATA_MAIL)], sizeof(struct MailStruct));
```
Comment (PikalaxALT) [inline src/mail_data.c]: If you do `gSaveBlock1Ptr->mail[i] = gSaveBlockPtr->mail[GetMonData(mon, MON_DATA_MAIL)];`, does it still match?

File: src/fldeff_dig.c (hunk 17 above)
Comment (jiangzhengwenjz) [reply src/fldeff_dig.c]: Actually all this kind of name has counterpart in pokeem. Should I still document them with my opinion?

File: src/mail_data.c (hunk 24 above)
Comment (jiangzhengwenjz) [reply src/mail_data.c]: It doesn't work. I can use species macro tho

File: src/mail_data.c (hunk 25 above)
Comment (jiangzhengwenjz) [reply src/mail_data.c]: No. I corrected `gSaveBlockPtr` to `gSaveBlock1Ptr` and it still doesn't match

File: src/fldeff_dig.c (hunk 17 above)
Comment (PikalaxALT) [reply src/fldeff_dig.c]: Whatever you decide, but the `hm2_` names simply will not fly.

File: src/mail_data.c (hunk 24 above)
Comment (PikalaxALT) [reply src/mail_data.c]: Try 30000 instead of 1875, I miscounted the 0s.

Comment (jiangzhengwenjz) [reply src/mail_data.c]: Still no luck. They actually optimized the function in EM is the word 0x7530 is used instead of 0x7530 << 16

Comment (PikalaxALT) [reply src/mail_data.c]:
```c

u16 SpeciesToMailSpecies(u16 species, u32 personality)
{
    if (species == SPECIES_UNOWN) {
        u32 mailSpecies = GetUnownLetterByPersonality(personality) + 30000;
        return mailSpecies;
    }
    return species;
}
```

File: src/fldeff_dig.c (hunk 18 above)
Comment (jiangzhengwenjz) [reply src/fldeff_dig.c]: Do you mean I should define a macro called `FLDEFF_SET_FUNCTION_TO_DATA(func)` or so?

Comment (PikalaxALT) [reply src/fldeff_dig.c]: Yes.

Comment (jiangzhengwenjz) [reply src/fldeff_dig.c]: do you expect I to add a semicolon when using the macro so that it looks like a function?

Comment (PikalaxALT) [reply src/fldeff_dig.c]: Yes, as it is stylistically consistent with the remainder of the repository.

**Pull request merged**